### DISCUSSION TITLE
Migrations - changing key types from Integer to Text (UUID)

### DISCUSSION THREAD

**Discussion by JanezStupar** [simolus3/drift#2723] [Q&A] [answered]
My app has two tables. 

I am looking to change primary keys from - incremental  Integer id's to text UUID's.

I would like to keep the `id` column as primary key.

I also have a reference from first table to the second (without FK constraint, because that didn't exist when I started working on my app).

My problem is that in the Drift documentation it is not clear to me how I should go about writing a migration where I change the type of the column. Generate a new value and update the existing references.

**Comment by simolus3** [comment, marked as answer]
> I am looking to change primary keys from - incremental Integer id's to text UUID's.

This is generally possible with the [alterTable](https://drift.simonbinder.eu/docs/migrations/api/#complex-migrations) API running [this procedure](https://www.sqlite.org/lang_altertable.html#otheralter) for you.

sqlite will let you copy the existing integers into the new text column, but at that point they're not valid UUIDs obviously. If you have a deterministic mechanism of converting the integer ids to uuids, you can use that as a `columnTransformer` in the `alterTable` call. You will also need to update the same table in a similar way. If you want to generate random uuids for values in the table, I would first select the ids from the existing table, create a map of ids to uuids, re-create the table and update everything based on that map.

**Comment by JanezStupar** [comment]
>
> If you want to generate random uuids for values in the table, I would
> first select the ids from the existing table, create a map of ids to uuids,
> re-create the table and update everything based on that map..

Yes this would be my preferred approach. What I am wondering is how to
express that in the context of a Migration.

~~How can I access the mapping inside the transformer and make sure it gets
applied to the new table?~~

Nvm, I figured it out. Will post my solution for reference when I wrap it up.

**Reply by JanezStupar** [reply]
Okay, just for the completeness sake - if any other poor soul comes looking.

This is how I got it to work:
```
class MigrationFrom1To2 {
  static Future<void> onUpgrade(Migrator m, schema) async {
    await m.alterTable(TableMigration(schema.epochs, columnTransformer: {
      schema.epochs.id: schema.epochs.id.cast<String>(),
    }));
    await m.alterTable(TableMigration(schema.categories, columnTransformer: {
      schema.categories.id: schema.categories.id.cast<String>(),
    }));
  }

  static Future<void> onBeforeOpen(Database db, OpeningDetails details) async {
    final Map _migrationCache = {};
    _migrationCache['epochs'] = {};
    _migrationCache['categories'] = {'1': Value('')};
    var epochsCache = _migrationCache['epochs'];
    var categoriesCache = _migrationCache['categories'];
    return await queryDeleted(db.executor, () async {
      await db.transaction(() async {
        final categoriesRows = await db.select(db.categories).get();
        final epochRows = await db.select(db.epochs).get();

        await db.batch((b) {
          categoriesRows.forEach((category) {
            if (categoriesCache[category.id] == null) {
              categoriesCache[category.id] = ShortUuid().generate();
            }
            b.update(db.categories,
                category.copyWith(id: categoriesCache[category.id]),
                where: (tbl) => tbl.id.equals(category.id));
          });
          epochRows.forEach((epoch) {
            if (epochsCache[epoch.id] == null) {
              epochsCache[epoch.id] = ShortUuid().generate();
            }
            (b.update(
              db.epochs,
              epoch.copyWith(
                  id: epochsCache[epoch.id],
                  category: categoriesCache[epoch.category ?? '']),
              where: (tbl) => tbl.id.equals(epoch.id),
            ));
          });
        });
      });
    });
  }
}

```
